File: _mkdocs/javascripts/widgets/living-capital.js
/*
 * «Сколько нужно, чтобы жить с трейдинга» — холодный душ в местных цифрах.
 *
 * Зачем: «брось работу и живи с трейдинга» — обещание, на котором продают
 * курсы и сигналы. Проект честно говорит, что реалистичная доходность скучная,
 * но нигде не показывал, что из этого следует в деньгах. А следует простое:
 * капитал должен быть примерно в семьдесят раз больше месячных расходов.
 *
 * Математика — зеркало forex_toolkit/living_capital.py, сверка в tests_e2e.
 */
(function () {
  var root = document.getElementById("living-capital-widget");
  if (!root || !window.FXW) return;

  var F = window.FXW;
  var TAX_RATE = 0.12;
  var MAX_MONTHS = 12 * 60;

  var T = F.pick({
    ru: {
      needLabel: "Сколько нужно в месяц на жизнь, $",
      returnLabel: "Месячная доходность, %",
      bufferLabel: "Подушка, месяцев расходов",
      startLabel: "Есть сейчас, $",
      addLabel: "Могу откладывать в месяц, $",
      rateLabel: "Курс USD → UZS",
      calc: "Посчитать",
      capital: "Торговый капитал",
      buffer: "Подушка отдельно",
      total: "Всего нужно",
      gross: "Заработать до налога",
      time: "Копить при твоих вводных",
      never: "при таких вводных цель не достигается",
      months: function (n) {
        var years = Math.floor(n / 12), rest = n % 12;
        return years ? years + " г. " + rest + " мес." : n + " мес.";
      },
      ratio: function (times) {
        return "Капитал больше месячных расходов в " + times + " раз.";
      },
      note: "Расчёт учитывает НДФЛ 12%: чтобы получить на руки, заработать надо больше. Подушку держат отдельно от торгового счёта — иначе убыточный месяц съедает сам капитал.",
      warn: "Это оценка порядка величины при заданной доходности, а не план. Стабильной доходности не бывает: убыточные месяцы есть у всех, и именно поэтому снимать всё до копейки нельзя.",
      errNeed: "Укажи месячные расходы больше нуля.",
      errReturn: "Месячная доходность — от 0.1% до 99%.",
      errRate: "Курс должен быть больше нуля."
    },
    en: {
      needLabel: "Monthly living costs, $",
      returnLabel: "Monthly return, %",
      bufferLabel: "Buffer, months of expenses",
      startLabel: "Have right now, $",
      addLabel: "Can save per month, $",
      rateLabel: "USD → UZS rate",
      calc: "Calculate",
      capital: "Trading capital",
      buffer: "Buffer, held separately",
      total: "Needed in total",
      gross: "Must earn before tax",
      time: "Time to save it up",
      never: "not reachable with these inputs",
      months: function (n) {
        var years = Math.floor(n / 12), rest = n % 12;
        return years ? years + "y " + rest + "m" : n + " months";
      },
      ratio: function (times) {
        return "The capital is " + times + " times your monthly costs.";
      },
      note: "The figure includes 12% income tax: to take money home you must earn more. The buffer is held away from the trading account — otherwise a losing month eats the capital itself.",
      warn: "This is an order-of-magnitude estimate at the return you entered, not a plan. Returns are never steady: everyone has losing months, which is exactly why you cannot withdraw every last dollar.",
      errNeed: "Enter monthly costs above zero.",
      errReturn: "Monthly return must be between 0.1% and 99%.",
      errRate: "The rate must be above zero."
    },
    uz: {
      needLabel: "Oyiga yashash uchun kerak, $",
      returnLabel: "Oylik daromad, %",
      bufferLabel: "Zaxira, necha oylik xarajat",
      startLabel: "Hozir bor, $",
      addLabel: "Oyiga jamg'ara olaman, $",
      rateLabel: "USD → UZS kursi",
      calc: "Hisoblash",
      capital: "Savdo kapitali",
      buffer: "Zaxira, alohida saqlanadi",
      total: "Jami kerak",
      gross: "Soliqqacha ishlash kerak",
      time: "Shu vvodnilar bilan jamg'arish",
      never: "bunday shartlarda maqsadga erishilmaydi",
      months: function (n) {
        var years = Math.floor(n / 12), rest = n % 12;
        return years ? years + " yil " + rest + " oy" : n + " oy";
      },
      ratio: function (times) {
        return "Kapital oylik xarajatdan " + times + " barobar katta.";
      },
      note: "Hisobda JShDS 12% inobatga olingan: qo'lga olish uchun ko'proq ishlash kerak. Zaxira savdo hisobidan alohida saqlanadi — aks holda zararli oy kapitalning o'zini yeydi.",
      warn: "Bu siz kiritgan daromad bo'yicha kattalik tartibi bahosi, reja emas. Barqaror daromad bo'lmaydi: zararli oylar hammada bor, shuning uchun hammasini oxirgi tiyinigacha yechib bo'lmaydi.",
      errNeed: "Oylik xarajatni noldan katta kiriting.",
      errReturn: "Oylik daromad 0.1% dan 99% gacha bo'lishi kerak.",
      errRate: "Kurs noldan katta bo'lishi kerak."
    }
  });

  function monthsToReach(target, start, monthlyAdd, monthlyReturn) {
    if (target <= 0) return 0;
    if (start >= target) return 0;
    if (monthlyReturn <= 0 && monthlyAdd <= 0) return null;
    var balance = start;
    for (var month = 1; month <= MAX_MONTHS; month++) {
      balance = balance * (1 + monthlyReturn) + monthlyAdd;
      if (balance >= target) return month;
    }
    return null;
  }

  function planFor(need, monthlyReturn, bufferMonths, start, monthlyAdd) {
    var gross = need / (1 - TAX_RATE);
    var capital = gross / monthlyReturn;
    var buffer = need * bufferMonths;
    var total = capital + buffer;
    return {
      monthly_need: need,
      gross_needed: gross,
      required_capital: capital,
      buffer: buffer,
      total_needed: total,
      months_to_reach: monthsToReach(total, start, monthlyAdd, monthlyReturn)
    };
  }

  window.__fxLivingCapital = planFor;

  root.innerHTML = '<div class="pos-calc-widget">' +
    field("lc-need", T.needLabel, 500, 1, "any") +
    field("lc-return", T.returnLabel, 1.5, 0.1, "0.1") +
    field("lc-buffer", T.bufferLabel, 6, 0, "1") +
    field("lc-start", T.startLabel, 1000, 0, "any") +
    field("lc-add", T.addLabel, 200, 0, "any") +
    field("lc-rate", T.rateLabel, 12500, 1, "any") +
    '<button type="button" id="lc-calc-btn" class="pc-row-wide">' + F.escape(T.calc) + "</button>" +
    '<div id="lc-result" role="status" aria-live="polite" hidden></div></div>';

  function field(id, label, value, min, step) {
    return '<label for="' + id + '">' + F.escape(label) + "</label>" +
      '<input type="number" id="' + id + '" value="' + value + '" min="' + min +
      '" step="' + step + '" autocomplete="off">';
  }

  function uzs(value, rate) {
    return F.int(Math.round(value * rate)) + " UZS";
  }

  // F.money жёстко ставит minimumFractionDigits: 2, поэтому просить у неё
  // ноль знаков нельзя — Intl бросает RangeError, когда минимум больше
  // максимума. Для крупных сумм округляем сами.
  function whole(value) {
    return "$" + F.int(Math.round(value));
  }

  function row(label, value, extra) {
    return '<div class="tax-result-row"><span>' + F.escape(label) + "</span><span>" +
      value + (extra ? ' <small>' + extra + "</small>" : "") + "</span></div>";
  }

  function calculate() {
    var need = parseFloat(document.getElementById("lc-need").value);
    var percent = parseFloat(document.getElementById("lc-return").value);
    var bufferMonths = parseInt(document.getElementById("lc-buffer").value, 10);
    var start = parseFloat(document.getElementById("lc-start").value);
    var add = parseFloat(document.getElementById("lc-add").value);
    var rate = parseFloat(document.getElementById("lc-rate").value);
    var box = document.getElementById("lc-result");

    var errors = [];
    if (!(need > 0)) errors.push(T.errNeed);
    if (!(percent >= 0.1 && percent <= 99)) errors.push(T.errReturn);
    if (!(rate > 0)) errors.push(T.errRate);
    if (!(bufferMonths >= 0)) bufferMonths = 0;
    if (!(start >= 0)) start = 0;
    if (!(add >= 0)) add = 0;

    box.hidden = false;
    if (errors.length) {
      box.className = "calc-result calc-error";
      box.innerHTML = errors.map(function (message) {
        return "<p>⛔ " + F.escape(message) + "</p>";
      }).join("");
      return;
    }

    var plan = planFor(need, percent / 100, bufferMonths, start, add);
    var times = Math.round(plan.total_needed / need);

    box.className = "calc-result calc-warn";
    box.innerHTML =
      '<div class="tax-headline">' + whole(plan.total_needed) + "</div>" +
      '<div class="tax-subhead">' + F.escape(T.total) + "</div>" +
      row(T.capital, whole(plan.required_capital), uzs(plan.required_capital, rate)) +
      row(T.buffer, whole(plan.buffer), uzs(plan.buffer, rate)) +
      row(T.gross, F.money(plan.gross_needed, 2)) +
      row(T.time, plan.months_to_reach === null
        ? F.escape(T.never)
        : F.escape(T.months(plan.months_to_reach))) +
      "<p><strong>" + F.escape(T.ratio(times)) + "</strong></p>" +
      '<p class="journal-tax-note">' + F.escape(T.note) + "</p>" +
      '<p class="journal-tax-note">' + F.escape(T.warn) + "</p>";

    if (window.fxTrack) window.fxTrack("calculator_completed");
  }

  document.getElementById("lc-calc-btn").addEventListener("click", calculate);
  ["lc-need", "lc-return", "lc-buffer", "lc-start", "lc-add", "lc-rate"].forEach(function (id) {
    document.getElementById(id).addEventListener("change", calculate);
  });
  calculate();
})();

File: forex_toolkit/living_capital.py
"""Сколько нужно капитала, чтобы жить с трейдинга.

Зачем: «бросить работу и жить с трейдинга» — обещание, на котором продают
курсы и «сигналы». Проект честно говорит, что реалистичная доходность
скучная, но нигде не показывает, что из этого следует **в деньгах**. А следует
простое: чтобы снимать сумму месячных расходов, капитал должен быть примерно в
семьдесят раз больше неё. Увидев это число, человек принимает другое решение,
чем услышав «торговля — это не быстрые деньги».

Расчёт::

    снять на руки  ->  заработать до налога  ->  какой капитал это даёт
    need           ->  need / (1 - tax)      ->  gross / monthly_return

Что учтено и почему:

* **налог**: чтобы получить на руки, заработать надо больше на ставку НДФЛ;
* **подушка**: если снимать всю прибыль каждый месяц, убыточный месяц придётся
  покрывать из капитала. Поэтому рядом считается запас расходов на N месяцев,
  который держат отдельно от торгового счёта;
* **срок**: за сколько месяцев такой капитал накопится при текущем депозите и
  ежемесячном пополнении.

Чего расчёт НЕ обещает: что доходность будет стабильной. Это оценка порядка
величины при заданной ставке, а не план. Убыточные месяцы бывают у всех, и
именно поэтому снимать всё до копейки нельзя.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

# Ставка НДФЛ резидента Узбекистана — та же, что в tax_summary.
DEFAULT_TAX_RATE = 0.12

# Сколько месяцев расходов держать вне торгового счёта. Три месяца — не
# норматив, а нижняя граница здравого смысла: типичная серия убытков короче.
DEFAULT_BUFFER_MONTHS = 6

# Дальше этого горизонта прогноз бессмысленен, а цикл должен завершаться.
MAX_MONTHS = 12 * 60


@dataclass(frozen=True)
class LivingPlan:
    monthly_need: float
    monthly_return: float
    tax_rate: float
    gross_needed: float
    required_capital: float
    buffer: float
    total_needed: float
    months_to_reach: int | None
    """``None``, если при заданных вводных цель не достигается за 60 лет."""

    def as_dict(self) -> dict[str, float | int | None]:
        return {
            "monthly_need": round(self.monthly_need, 2),
            "gross_needed": round(self.gross_needed, 2),
            "required_capital": round(self.required_capital, 2),
            "buffer": round(self.buffer, 2),
            "total_needed": round(self.total_needed, 2),
            "months_to_reach": self.months_to_reach,
        }


def months_to_reach(
    target: float, *, start: float, monthly_add: float, monthly_return: float
) -> int | None:
    """За сколько месяцев капитал дорастёт до цели.

    Считаем шагами, а не формулой: пополнение и доходность применяются в том же
    порядке, что в жизни, и функция честно возвращает ``None``, если цель
    недостижима (например, пополнения нет, а доходность нулевая).
    """
    if target <= 0:
        return 0
    if start >= target:
        return 0
    if monthly_return <= 0 and monthly_add <= 0:
        return None

    balance = start
    for month in range(1, MAX_MONTHS + 1):
        balance = balance * (1 + monthly_return) + monthly_add
        if balance >= target:
            return month
    return None


def plan_for(
    *,
    monthly_need: float,
    monthly_return: float,
    tax_rate: float = DEFAULT_TAX_RATE,
    buffer_months: int = DEFAULT_BUFFER_MONTHS,
    start: float = 0.0,
    monthly_add: float = 0.0,
) -> LivingPlan:
    """Капитал, подушка и срок для заданных месячных расходов."""
    if monthly_need <= 0:
        raise ValueError("месячные расходы должны быть положительными")
    if not 0 < monthly_return < 1:
        raise ValueError("месячная доходность задаётся долей от 0 до 1")
    if not 0 <= tax_rate < 1:
        raise ValueError("ставка налога задаётся долей от 0 до 1")
    if buffer_months < 0:
        raise ValueError("подушка не может быть отрицательной")
    if not math.isfinite(start) or not math.isfinite(monthly_add):
        raise ValueError("стартовый капитал и пополнение должны быть числами")

    gross = monthly_need / (1 - tax_rate)
    capital = gross / monthly_return
    buffer = monthly_need * buffer_months
    total = capital + buffer
    return LivingPlan(
        monthly_need=monthly_need,
        monthly_return=monthly_return,
        tax_rate=tax_rate,
        gross_needed=gross,
        required_capital=capital,
        buffer=buffer,
        total_needed=total,
        months_to_reach=months_to_reach(
            total, start=start, monthly_add=monthly_add, monthly_return=monthly_return
        ),
    )
